components/script/dom/htmlimageelement.rs

```diff
@@ -782,6 +797,71 @@ impl LayoutHTMLImageElementHelpers for LayoutJS<HTMLImageElement> {
     }
 }
 
+pub fn parse_a_sizes_attribute(input: DOMString, width: Option<u32>) -> Result<Vec<Size>, String> {
+let mut sizes = Vec::<Size>::new();
+let unparsed_sizes = input.deref().split(',').collect::<Vec<_>>();
+
+    for unparsed_size in &unparsed_sizes {
+        let whitespace = unparsed_size.chars().rev().take_while(|c| char::is_whitespace(*c)).count();
+        let trimmed: String = unparsed_size.chars().take(unparsed_size.chars().count() - whitespace).collect();
+
+        if trimmed.is_empty() {
+            // return Err("reason".to_owned());
+            continue;
+        }
+        let mut input = ParserInput::new(&trimmed);
+        let url = ServoUrl::parse("about:blank").unwrap();
+        let reporter = NullReporter;
+        let context = ParserContext::new_for_cssom(&url,
+                                                        &reporter,
+                                                        None,
+                                                        ParsingMode::empty(),
+                                                        QuirksMode::NoQuirks);
+        let url = ServoUrl::parse("about:blank").unwrap();
+        let length = Parser::new(&mut input).try(|i| Length::parse_non_negative(&context, i));
+        match length {
+            Ok(len) => sizes.push(Size {
+                length: len,
+                query: None
+            }),
+            Err(_) => {
+                // return Err("reason".to_owned());
+                let mut media_query_parser = Parser::new(&mut input);
+                let media_query = media_query_parser.try(|i| MediaQuery::parse(&context, i));
+                match media_query {
+                    Ok(query) => {
+                        let length = media_query_parser.try(|i| Length::parse_non_negative(&context, i));
```

Comment:
We can use `Length::parse_non_negative` directly instead of `Parser::try`: `let length = Length::parse_non_negative(&context, &mut media_query_parser);`